src/throttler/throttler.ts

```diff
@@ -131,6 +145,37 @@ export abstract class Throttler<T, R> {
     return this.finishIfIdle();
   }
 
+  async handle(cursorIndex: number): Promise<void> {
+    const taskData = this.taskDataMap.get(cursorIndex);
+    if (!taskData) {
+      throw new Error(`taskData.get(${cursorIndex}) does not exist`);
+    }
+    switch (taskData.status) {
+      case TaskStatus.NOT_STARTED:
+        this.initializeTask(taskData, cursorIndex);
+        break;
+      case TaskStatus.RUNNING:
+        await this.executeTask(taskData, cursorIndex);
```

Comment:
If timeout is set to 2s, and the `this.handler(...)` call inside of `this.executeTask(...)` ends up taking 30s to run, this `await` here will force the caller to wait the whole 30s before rejecting the promise. I think there are 2 ways we can avoid that:

Option 1: Give the `setTimeout()` callback the ability to directly call the `reject(...)` function of a Promise in the task's promise chain. This is similar to [how I implemented a master timeout in the Admin SDK](https://github.com/FirebasePrivate/firebase-admin-node/blob/84cc27312b5fb5e531cf3526a653695ca032cceb/src/utils/api-request.ts#L923).

Option 2: Use `Promise.race([this.handler(...), setTimeoutPromise])`.

Reply:
Used Promise.race to race timeout promise and task execution promise